async-nats/src/connection.rs

```diff
@@ -40,254 +41,252 @@ pub(crate) struct Connection {
 /// Helds connection with NATS Server and communicates with `Client` via channels.
 impl Connection {
     pub(crate) fn try_read_op(&mut self) -> Result<Option<ServerOp>, io::Error> {
-        if self.buffer.starts_with(b"+OK\r\n") {
-            self.buffer.advance(5);
-            return Ok(Some(ServerOp::Ok));
+        let maybe_len = self.buffer.find(b"\r\n");
+        if maybe_len.is_none() {
+            return Ok(None);
         }
 
-        if self.buffer.starts_with(b"PING\r\n") {
-            self.buffer.advance(6);
+        let len = maybe_len.unwrap();
 
-            return Ok(Some(ServerOp::Ping));
+        if self.buffer.starts_with(b"+OK") {
+            self.buffer.advance(len + 2);
+            return Ok(Some(ServerOp::Ok));
         }
 
-        if self.buffer.starts_with(b"PONG\r\n") {
-            self.buffer.advance(6);
+        if self.buffer.starts_with(b"PING") {
+            self.buffer.advance(len + 2);
+            return Ok(Some(ServerOp::Ping));
+        }
 
+        if self.buffer.starts_with(b"PONG") {
+            self.buffer.advance(len + 2);
             return Ok(Some(ServerOp::Pong));
         }
 
         if self.buffer.starts_with(b"-ERR") {
-            if let Some(len) = self.buffer.find(b"\r\n") {
-                let line = std::str::from_utf8(&self.buffer[5..len])
-                    .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?;
-                let error_message = line.trim_matches('\'').to_string();
-                self.buffer.advance(len + 2);
+            let message = str::from_utf8(&self.buffer[5..len])
+                .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?
+                .trim_matches('\'')
+                .to_string();
 
-                return Ok(Some(ServerOp::Error(ServerError::new(error_message))));
-            }
-        }
-
-        if self.buffer.starts_with(b"INFO ") {
-            if let Some(len) = self.buffer.find(b"\r\n") {
-                let line = std::str::from_utf8(&self.buffer[5..len])
-                    .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?;
+            self.buffer.advance(len + 2);
 
-                let server_info = serde_json::from_str(line)
-                    .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?;
+            return Ok(Some(ServerOp::Error(ServerError::new(message))));
+        }
 
-                self.buffer.advance(len + 2);
+        if self.buffer.starts_with(b"INFO") {
```

Comment:
I think you want space inside quotes `"INFO "` because `buffer[5..len]` below would accept non-space chars in that position

Reply:
Spot on.